### PR TITLE
Analyze and DB connection

### PR DESCRIPTION
**Pull request by yairvogel**

### CODE REVIEW

File: DbConnection.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+            await m_client.CreateDatabaseIfNotExistsAsync(new Database { Id = c_dbName });
+            var collectionDefinition = new DocumentCollection { Id = c_colName };
+            await m_client.CreateDocumentCollectionIfNotExistsAsync(
+                UriFactory.CreateDatabaseUri(c_dbName),
+                collectionDefinition);
+        }
+
+        public static async Task AddAggregateDoc(AggregateRecord docDefinition)
+        {
+            using (m_client = new DocumentClient(new Uri(c_endPointUri), c_primaryKey))
```
Comment (shaybenh7) [inline DbConnection.cs:31]: initialize client once and use it

File: DbConnection.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+using System.Collections;
+using System.Configuration;
+using System.Threading.Tasks;
+using Microsoft.Azure.Documents;
+using Microsoft.Azure.Documents.Client;
+
+
+namespace OnBoardAssignment
+{
+    static class DbConnection
```
Comment (shaybenh7) [inline DbConnection.cs:11]: remove static.
rename class to CosmosDbClient and set as singelton.
initialize it once and use the same instance throughout your entire program

File: DbConnection.cs [hunk 3]
```diff
@@ -1,23 +1,53 @@
...
+using System.Diagnostics.Contracts;
 using System.Threading.Tasks;
 using Microsoft.Azure.Documents;
 using Microsoft.Azure.Documents.Client;
 
 
 namespace OnBoardAssignment
 {
-    static class DbConnection
+    class DbConnection : IDisposable
```
Comment (shaybenh7) [inline DbConnection.cs:13]: public

File: DbConnection.cs [hunk 4]
```diff
@@ -1,23 +1,53 @@
...
+        private readonly string c_dbName = "testDB";
+        private string c_colName = "testCollection";
+        private DocumentClient m_client;
 
-        public static async Task CreateCollectionIfNotExists()
+        private static DbConnection? m_instance = null;
+        private static readonly object c_instanceLock = new object();
+        private bool m_disposed = false;
+
+        public DbConnection()
```
Comment (shaybenh7) [inline DbConnection.cs:26]: singelton pattern says ctor need to be private

File: DbConnection.cs [hunk 5]
```diff
@@ -1,23 +1,53 @@
...
+        {
+            m_client = new DocumentClient(new Uri(c_endPointUri), c_primaryKey );
+        }
+
+        public static DbConnection Instance
+        {
+            get
+            {
+                lock (c_instanceLock)
```
Comment (shaybenh7) [inline DbConnection.cs:35]: we are doing doing a concurrent usage of the dbconnction, therefore no need to lock

File: DbConnection.cs [hunk 6]
```diff
@@ -26,24 +56,29 @@ public static async Task CreateCollectionIfNotExists()
...
-        public static async Task AddAggregateDoc(AggregateRecord docDefinition)
+        public async Task AddAggregateDoc(AggregateRecord docDefinition)
         {
-            using (m_client = new DocumentClient(new Uri(c_endPointUri), c_primaryKey))
-            {
-                await CreateCollectionIfNotExists();
-                Uri colUri = UriFactory.CreateDocumentCollectionUri(c_dbName, c_colName);
-                await m_client.CreateDocumentAsync(colUri, docDefinition);
-            }
+            Uri colUri = UriFactory.CreateDocumentCollectionUri(c_dbName, c_colName);
```
Comment (shaybenh7) [inline DbConnection.cs:61]: what does this function doing when the collection already exists?

File: DbConnection.cs [hunk 7]
```diff
@@ -26,24 +56,29 @@ public static async Task CreateCollectionIfNotExists()
...
+            await m_client.CreateDocumentAsync(colUri, docDefinition);
```
Comment (shaybenh7) [inline DbConnection.cs:62]: await m_client.CreateDocumentAsync(colUri, docDefinition).ConfigureAwait(false);

File: DbConnection.cs [hunk 8]
```diff
@@ -26,24 +56,29 @@ public static async Task CreateCollectionIfNotExists()
...
         }
 
-        public static IEnumerable GetAggregateRecords()
+        public IEnumerable GetAggregateRecords()
         {
-            using (m_client = new DocumentClient(new Uri(c_endPointUri), c_primaryKey))
+            Uri colUri = UriFactory.CreateDocumentCollectionUri(c_dbName, c_colName);
+            return m_client.CreateDocumentQuery(colUri, 
+                "select c.Id, c.EquationType, c.OperatorType from c");
```
Comment (shaybenh7) [inline DbConnection.cs:69]: extract query to const members

File: DbConnection.cs (hunk 3 above)
Comment (yairvogel) [reply DbConnection.cs:12]: done

File: DbConnection.cs (hunk 4 above)
Comment (yairvogel) [reply DbConnection.cs:25]: done

File: DbConnection.cs (hunk 5 above)
Comment (yairvogel) [reply DbConnection.cs]: done

File: DbConnection.cs (hunk 8 above)
Comment (yairvogel) [reply DbConnection.cs:57]: are you referring to the query string?

**Pull request merged**